File: app/src/main/java/com/example/recipeclipper/data/DefaultListRepository.kt
package com.example.recipeclipper.data

import com.example.recipeclipper.data.local.dao.ListDao
import com.example.recipeclipper.data.local.entity.RecipeListCrossRef
import com.example.recipeclipper.data.model.RecipeList
import com.example.recipeclipper.data.model.RecipeSummary
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import javax.inject.Inject
import javax.inject.Singleton

/**
 * The real, Room-backed [ListRepository]. Bound to the interface with `@Binds`.
 *
 * Database failures are logged and degrade instead of crashing (see [ErrorLog.guard]): the
 * writes become no-ops, a Flow emits an empty list, and [createList] returns [CREATE_FAILED].
 */
@Singleton
class DefaultListRepository @Inject constructor(
    private val listDao: ListDao,
    private val clock: Clock,
    private val log: ErrorLog
) : ListRepository {

    override fun observeLists(): Flow<List<RecipeList>> =
        listDao.observeLists(ListDao.NO_RECIPE).map { rows -> rows.map { it.toDomain() } }
            .orEmptyOnError(log, "observeLists")

    override fun observeListsFor(recipeId: Long): Flow<List<RecipeList>> =
        listDao.observeLists(recipeId).map { rows -> rows.map { it.toDomain() } }
            .orEmptyOnError(log, "observeListsFor")

    override fun observeRecipesIn(listId: Long): Flow<List<RecipeSummary>> =
        listDao.observeRecipesIn(listId).map { rows -> rows.map { it.toDomain() } }
            .orEmptyOnError(log, "observeRecipesIn")

    override suspend fun setMembership(recipeId: Long, listId: Long, inList: Boolean) =
        log.guard("setMembership", Unit) {
            if (inList) {
                listDao.addToList(RecipeListCrossRef(recipeId, listId, clock.now()))
            } else {
                listDao.removeFromList(recipeId, listId)
            }
        }

    override suspend fun createList(name: String, addRecipeId: Long?): Long =
        log.guard("createList", CREATE_FAILED) {
            listDao.create(name.trim(), addRecipeId ?: ListDao.NO_RECIPE, clock.now())
        }

    override suspend fun rename(listId: Long, name: String) =
        log.guard("rename", Unit) { listDao.rename(listId, name.trim()) }

    override suspend fun deleteList(listId: Long) =
        log.guard("deleteList", Unit) { listDao.delete(listId) }

    companion object {
        /** What [createList] returns when the insert failed: an id no row can have (Room ids
         *  start at 1). No caller uses the returned id today; one that does must check it. */
        const val CREATE_FAILED = -1L
    }
}

File: ios/RecipeClipper/Data/Local/DefaultListRepository.swift
import Combine
import Foundation

/// The real, SQLite-backed ListRepository (Android's DefaultListRepository). Database failures
/// are logged; the contract's methods don't throw.
final class DefaultListRepository: ListRepository {
    private let db: AppDatabase
    private let clock: Clock

    init(db: AppDatabase, clock: Clock) {
        self.db = db
        self.clock = clock
    }

    func observeLists() -> AnyPublisher<[RecipeList], Never> {
        observeLists(recipeId: ListDao.noRecipe)
    }

    func observeListsFor(recipeId: Int64) -> AnyPublisher<[RecipeList], Never> {
        observeLists(recipeId: recipeId)
    }

    private func observeLists(recipeId: Int64) -> AnyPublisher<[RecipeList], Never> {
        db.observe { conn in try ListDao(db: conn).lists(recipeId: recipeId).map { $0.toDomain() } }
    }

    func observeRecipesIn(listId: Int64) -> AnyPublisher<[RecipeSummary], Never> {
        db.observe { conn in try ListDao(db: conn).recipesIn(listId: listId).map { $0.toDomain() } }
    }

    func setMembership(recipeId: Int64, listId: Int64, inList: Bool) async {
        let now = clock.now()
        await perform("setMembership") { dao in
            if inList {
                try dao.addToList(ListMembership(recipeId: recipeId, listId: listId, addedAt: now))
            } else {
                try dao.removeFromList(recipeId: recipeId, listId: listId)
            }
        }
    }

    @discardableResult
    func createList(name: String, addRecipeId: Int64?) async -> Int64 {
        let now = clock.now()
        let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
        do {
            return try await db.write { conn in
                try ListDao(db: conn).create(name: trimmed, recipeId: addRecipeId ?? ListDao.noRecipe, now: now)
            }
        } catch {
            dataLog.error("createList failed: \(String(describing: error), privacy: .public)")
            return 0
        }
    }

    func rename(listId: Int64, name: String) async {
        let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
        await perform("rename") { dao in try dao.rename(id: listId, name: trimmed) }
    }

    func deleteList(listId: Int64) async {
        await perform("deleteList") { dao in try dao.delete(id: listId) }
    }

    private func perform(_ what: String, _ body: @escaping (ListDao) throws -> Void) async {
        do {
            try await db.write { conn in try body(ListDao(db: conn)) }
        } catch {
            dataLog.error("\(what, privacy: .public) failed: \(String(describing: error), privacy: .public)")
        }
    }
}
